src/qz/ws/PrintSocketClient.java

```diff
@@ -461,6 +475,97 @@ private void processMessage(Session session, JSONObject json, SocketConnection c
                 break;
             }
 
+            case FILE_START_LISTENING: {
+                FileParams fileParams = new FileParams(params);
+                Path absPath = FileUtilities.getAbsolutePath(params, shownCertificate, true);
+                FileIO fileIO = new FileIO(session, params, fileParams.getPath(), absPath);
+
+                if (connection.getFileListener(absPath) == null && !fileIO.isWatching()) {
+                    connection.addFileListener(absPath, fileIO);
+
+                    FileUtilities.setupListener(fileIO);
+                    sendResult(session, UID, null);
+                } else {
+                    sendError(session, UID, "Already listening to path events");
+                }
+
+                break;
+            }
+            case FILE_STOP_LISTENING: {
+                if (params.isNull("path")) {
+                    connection.removeAllFileListeners();
+                    sendResult(session, UID, null);
+                } else {
+                    FileParams fileParams = new FileParams(params);
+                    Path absPath = FileUtilities.createAbsolutePath(fileParams, shownCertificate);
+                    FileIO fileIO = connection.getFileListener(absPath);
+
+                    if (fileIO != null) {
+                        fileIO.close();
+                        FileWatcher.deregisterWatch(fileIO);
+                        connection.removeFileListener(absPath);
+                        sendResult(session, UID, null);
+                    } else {
+                        sendError(session, UID, "Not already listening to path events");
+                    }
+                }
+
+                break;
+            }
+            case FILE_LIST: {
+                Path absPath = FileUtilities.getAbsolutePath(params, shownCertificate, true);
+
+                if (Files.exists(absPath)) {
+                    if (Files.isDirectory(absPath)) {
+                        ArrayList<String> files = new ArrayList<>();
+                        Files.list(absPath).forEach(file -> files.add(file.getFileName().toString()));
+                        sendResult(session, UID, new JSONArray(files));
+                    } else {
+                        sendError(session, UID, "Path is not a directory");
+                    }
+                } else {
+                    sendError(session, UID, "Path does not exist");
```

Comment:
Can we do the equivalent of trying a `mkdir` here and then loop the call back onto itself?  That we aren't throwing errors right off the start.

Reply:
This call is just listing the files though. Do we really want a write operation on that?

Reply:
My concern is that the default folder isn't available and throws a false error and that's non-sane defaults.  Perhaps we swallow the exception if it's the root-most directory and send back a black list of files?  I really don't like someone clicking that "list files" and have the software yell at them about a folder they didn't know should exist.  It seems unintuitive to need to rely on the file write to get the parent folder.  Thoughts are welcome here.

Reply:
You're right, I forgot about the default folder on clean runs.
I've altered the code so that if we are performing a fileio operation, it will make sure at least that folder always exists first